src/index_intf.ml

```diff
@@ -134,7 +134,14 @@ module type S = sig
     t
   (** The constructor for indexes.
 
-      @param auto_flush_callback adds a callback before an auto flush.
+      @param auto_flush_callback A function to be called before any new bindings
+      are persisted to disk (including both automatic flushing and explicit
+      calls to {!flush}). Values flushed during {!close} are excluded.
```

Comment:
What do you think of not calling the callback on explicit calls to flush? In irmin, we always call the callback explicitly when calling index.flush https://github.com/mirage/irmin/blob/master/src/irmin-pack/pack.ml#L195

Reply:
There are two arguments _against_:

- Not calling the flush callback for explicit flushing requires the user to remember to do so (and presumably they will always have to do so if this is being used to maintain an invariant).
- It's convenient for testing if we have a way to trigger a flush in `log_async`, and without this we can only expose it by writing `1_000_000` bytes during a merge.

Arguments _for_ include:

- In cases of implementing an explicit flushing mechanism on top of this one (as in pack), it's perhaps more intuitive if the parent flushing is _not_ done via a callback.
-  `**auto**_flush_callback` is named to suggest that explicit flushes don't trigger it.

I see two solutions: (a) add some sort of other hook to fiddle with the `auto_flush_limit` so that we can test this properly, or (b) expose a switch on `Index.flush` (perhaps only in the private API for testing) that gives the user control over this behaviour.

What do you think?

Reply:
Indeed, that's what the name of the argument suggests, that requires to remove the `no_callback` argument in `flush` too.

Either that, or just name it `flush_callback`, which I tend to like more because it avoids the callback (as in https://github.com/mirage/irmin/blob/master/src/irmin-pack/pack.ml#L195). I can't see a reason why someone should have a different handling of automatic and manual calls to `flush`.

Reply:
I agree that `auto` strongly suggests that the callback is not called for explicit `index.flush` calls, and it makes sense to exclude both calls to `index.flush` and `index.close`.
And I like solution (b), seems simpler, but you can choose which one you prefer :)